[CodeCompletion] Migrate ForEachSequence and PostfixExprBeginning to solver-based

ahoppen opened the pull request:
Migrate two more completion kinds (ForEachSequence and PostfixExprBeginning) to the solver-based implementation.

ahoppen committed:
[CodeCompletion] Migrate ForEachSequence and PostfixExprBeginning to solver-based

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[Parser] Ignore generic arguments after a code completion token

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

xedin commented on lib/IDE/TypeCheckCompletionCallback.cpp:
```diff
@@ -92,18 +93,93 @@ Type swift::ide::getTypeForCompletion(const constraints::Solution &S, Expr *E) {
+/// \endcode
+/// If the code completion expression occurs in such an AST, return the
+/// declaration of the \c $match variable, otherwise return \c nullptr.
+static VarDecl *getMatchVarIfInPatternMatch(Expr *E, ConstraintSystem &CS) {
```
Good news is that `ConstraintSystem` already has a way to check this `ConstraintSystem::isPatternMatchingOperator`

ahoppen replied:
Nice. Thanks for catching this. Surely removes some ugly code.

ahoppen committed:
[CodeCompletion] Don't record solution results twice in ExprCompletion

This fixes a performance problem where we were sometimes reporting >1000 solutions but all the relevant parameters of the result were equal.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test
